### ISSUE TITLE
Deadlock using nested BatchDataLoader in v12

### ISSUE THREAD

**Issue by GraemeF** [ChilliCream/hotchocolate#4280]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

A `BatchDataLoader` that uses a nested `BatchDataLoader` results in a deadlock in v12, whereas it would work OK in v11.

### Steps to reproduce

1. Create a server
2. Add these types to the project:
```csharp
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Threading;
using System.Threading.Tasks;
using GreenDonut;
using HotChocolate;
using HotChocolate.Resolvers;
using HotChocolate.Types;
using HotChocolate.Types.Relay;

namespace Repro
{
    [ExtendObjectType("Query")]
    public class FooQueries
    {
        public async Task<FooObject?> GetFoo(IResolverContext context, CancellationToken ct) =>
            await FooObject.Get(context, "hello", ct);
    }

    [GraphQLName("Foo")]
    [Node]
    public class FooObject
    {
        public FooObject(string field)
        {
            id = field;
        }

        public string id { get; }
        public string field => id;

        public static async Task<FooObject?> Get(IResolverContext context, string id,
            CancellationToken ct) =>
            new((await context.DataLoader<FooDataLoader>().LoadAsync(id, ct)).Field);
    }

    public class FooDataLoader : BatchDataLoader<string, FooRecord>
    {
        private readonly FooNestedDataLoader _nestedDataLoader;

        public FooDataLoader(IBatchScheduler batchScheduler, [DataLoader] FooNestedDataLoader nestedDataLoader,
            DataLoaderOptions? options = null) : base(batchScheduler,
            options)
        {
            _nestedDataLoader = nestedDataLoader;
        }

        protected override async Task<IReadOnlyDictionary<string, FooRecord>> LoadBatchAsync(IReadOnlyList<string> keys,
            CancellationToken cancellationToken) =>
            (await _nestedDataLoader.LoadAsync(keys, cancellationToken)).ToImmutableDictionary(x => x.Field);
    }

    public class FooNestedDataLoader : BatchDataLoader<string, FooRecord>
    {
        public FooNestedDataLoader(IBatchScheduler batchScheduler, DataLoaderOptions? options = null) : base(
            batchScheduler,
            options)
        {
        }

        protected override async Task<IReadOnlyDictionary<string, FooRecord>> LoadBatchAsync(IReadOnlyList<string> keys,
            CancellationToken cancellationToken)
        {
            await Task.Delay(1, cancellationToken);
            return keys.ToImmutableDictionary(key => key, key => new FooRecord(key));
        }
    }

    public record FooRecord(string Field);
}
```
3. Register the dataloaders and queries object:
``` csharp
                .AddType<FooQueries>()
                .AddDataLoader<FooDataLoader>()
                .AddDataLoader<FooNestedDataLoader>()
```
4. Run this query:
``` graphql
query Foo { foo { id field } }
```

### Product

Hot Chocolate

### Version

12.0.0

**Comment by michaelstaib** [comment]
Its supported, this might be a regression that has slipped through.

**Comment by michaelstaib** [comment]
This bug is now fixed with 12.2.2-preview.1
